### PR TITLE
Move [Space|Subnet]Info into core/network Package

### PR DESCRIPTION
**Pull request by manadart**
## Description of change

This patch relocates the following from `network` to `core/network`:
- The `Id` type.
- The `GenerateVirtualMACAddress` function.
- The `SubnetInfo` type.
- The `SpaceInfo` type.

In addition, the `state.SubnetInfo` is removed, as is translation logic associated with it.

Most of the change is the edits/additions to package imports. It is expected that these will clean back up as more network concerns are moved into core.

A follow-up to this patch will address various TODOs for exposing more subnet fields (including for migration) and for storing the full slice of subnet availability zones.

Depends on in-flight patches. First change is: https://github.com/juju/juju/commit/d63c4108a9794eadc3e00f566c49c6782a205bbd

## QA steps

No functional changes. All affected packages have green unit test suites.

## Documentation changes

None.

## Bug reference

N/A

### CODE REVIEW

**Comment by manadart** [comment]
[Move network.SpaceInfo and network.SubnetInfo to core/network](https://trello.com/c/lNNZGdAR/1634-move-networkspaceinfo-and-networksubnetinfo-to-core-network)

File: juju/testing/utils.go [hunk 1]
```diff
@@ -84,11 +85,16 @@ func AddSubnetsWithTemplate(c *gc.C, st *state.State, numSubnets uint, infoTempl
 			return buf.String()
 		}
 
-		info.ProviderId = network.Id(permute(string(info.ProviderId)))
+		info.ProviderId = corenetwork.Id(permute(string(info.ProviderId)))
 		info.CIDR = permute(info.CIDR)
-		info.AvailabilityZone = permute(info.AvailabilityZone)
 		info.SpaceName = permute(info.SpaceName)
 
+		zones := make([]string, len(info.AvailabilityZones))
```
Comment (manadart) [inline juju/testing/utils.go]: The new common type unifies availability zones as a slice where the state type was singular.

File: state/spacesdiscovery.go [hunk 2]
```diff
@@ -77,22 +78,12 @@ func (st *State) SaveSubnetsFromProvider(subnets []network.SubnetInfo, spaceName
...
-			ProviderId:        subnet.ProviderId,
-			ProviderNetworkId: subnet.ProviderNetworkId,
-			CIDR:              subnet.CIDR,
-			SpaceName:         spaceName,
-			VLANTag:           subnet.VLANTag,
-			AvailabilityZone:  firstZone,
-		})
+
+		subnet.SpaceName = spaceName
+		_, err = st.AddSubnet(subnet)
```
Comment (manadart) [inline state/spacesdiscovery.go]: This is a nice realisation of the change - we no longer convert `network.SubnetInfo` -> `state.SubnetInfo` -> `state.subnetDoc`.

File: core/network/subnet.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+	// FanInfo describes the fan networking setup for the subnet.
+	// It may be empty if this is not a fan subnet,
+	// or if this subnet information comes from a provider.
+	FanInfo *FanCIDRs
+}
+
+// AvailabilityZone returns the first of any availability zones,
+// or an empty string.
+func (s *SubnetInfo) AvailabilityZone() string {
```
Comment (manadart) [inline core/network/subnet.go]: Pushing this concern here means that we no longer have multiple blocks of the same logic to take a single AZ when translating for state storage. Once state stores a slice, this method can go.

File: core/network/subnet.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+	if len(s.AvailabilityZones) > 0 {
+		return s.AvailabilityZones[0]
```
Comment (hmlanigan) [inline core/network/subnet.go]: Can the order change?

File: upgrades/steps_27.go [hunk 5]
```diff
@@ -13,5 +13,12 @@ func stateStepsFor27() []Step {
 				return context.State().AddControllerNodeDocs()
 			},
 		},
+		&upgradeStep{
+			description: "recreated spaces with IDs",
```
Comment (hmlanigan) [inline upgrades/steps_27.go]: recreate?

File: state/spaces.go [hunk 6]
```diff
@@ -191,7 +238,7 @@ func (s *Space) EnsureDead() (err error) {
 
 	ops := []txn.Op{{
 		C:      spacesC,
-		Id:     s.doc.Name,
+		Id:     s.doc.Id,
```
Comment (hmlanigan) [inline state/spaces.go]: should this be DocID?

File: state/spaces.go [hunk 7]
```diff
@@ -247,3 +294,18 @@ func (s *Space) Refresh() error {
 	s.doc = doc
 	return nil
 }
+
+// createDefaultSpaceOp returns a transaction operation
+// that creates the default space (id=0).
+func createDefaultSpaceOp() txn.Op {
+	return txn.Op{
+		C:  spacesC,
+		Id: "0",
```
Comment (hmlanigan) [inline state/spaces.go]: st.docID("0")?

**Comment by manadart** [comment]
!!build!!

**Comment by manadart** [comment]
!!build!!

**Comment by manadart** [comment]
!!build!!

**Comment by manadart** [comment]
!!build!!

**Comment by manadart** [comment]
$$merge$$

**Pull request merged**